subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultTransformerInvoker.java

```diff
@@ -17,42 +17,386 @@
+
+        @Override
+        public Optional<Duration> getTimeout() {
+            return Optional.empty();
+        }
+
+        @Override
+        public void visitOutputs(OutputVisitor outputVisitor) {
+            outputVisitor.visitOutput("output", TreeType.DIRECTORY, ImmutableFileCollection.of(getOutputDir()));
+        }
+
+        @Override
+        public long markExecutionTime() {
+            // TODO Handle execution time
+            return 0;
+        }
+
+        @Override
+        public FileCollection getLocalState() {
+            return ImmutableFileCollection.of();
+        }
+
+        @Override
+        public void outputsRemovedAfterFailureToLoadFromCache() {
+        }
+
+        @Override
+        public CacheHandler createCacheHandler() {
+            return new CacheHandler() {
+                @Override
+                public <T> Optional<T> load(Function<BuildCacheKey, T> loader) {
+                    return Optional.empty();
+                }
+
+                @Override
+                public void store(Consumer<BuildCacheKey> storer) {
+                }
+            };
+        }
+
+        @Override
+        public void persistResult(ImmutableSortedMap<String, CurrentFileCollectionFingerprint> finalOutputs, boolean successful, OriginMetadata originMetadata) {
+            getResult().ifSuccessful(result -> {
+                historyRepository.persist(cacheKey.getPersistentCacheKey(), result, finalOutputs.values().iterator().next());
+            });
+        }
+
+        @Override
+        public Optional<ExecutionStateChanges> getChangesSincePreviousExecution() {
+            Optional<PreviousTransformerExecution> previousExecution = historyRepository.getPreviousExecution(cacheKey.getPersistentCacheKey());
+            return previousExecution.map(previous -> {
+                    CurrentFileCollectionFingerprint fingerprintBeforeExecution = outputFileCollectionFingerprinter.fingerprint(ImmutableFileCollection.of(getOutputDir()));
+                    ImmutableList.Builder<Change> builder = ImmutableList.builder();
+                    fingerprintBeforeExecution.visitChangesSince(previous.getOutputDirectoryFingerprint(), "outputDirectory", true, new ChangeVisitor() {
+                        @Override
+                        public boolean visitChange(Change change) {
+                            builder.add(change);
+                            return false;
+                        }
+                    });
+                    ImmutableList<Change> changes = builder.build();
+                    if (changes.isEmpty()) {
+                        this.result = previous.getResult();
+                    }
+                    return new TransformerExecutionStateChanges(changes);
+                }
+            );
+        }
+
+        @Override
+        public Optional<? extends Iterable<String>> getChangingOutputs() {
+            return Optional.of(ImmutableList.of(getOutputDir().getAbsolutePath()));
```

Comment:
Why not return `Optional<File>` here?

Reply:
We can return an `Optional<File>` for now, since it is only used for artifact transformations.

Reply:
Let's make this nicer later.